### PR TITLE
Let the Nextcloud-Container trust custom CAs (e.g. for using LDAPS) 2

### PR DESCRIPTION
**Pull request by lollo0296**
Signed-off-by: Lorenzo Marroccoli <lollo0296@gmail.com>

### CODE REVIEW

**Comment by lollo0296** [comment]
This is just a proposal / proof of concept which enables the user to make the `aio-nextcloud` container trust custom CA's.

# The reason for this PR
More info here: https://www.reddit.com/r/NextCloud/comments/wf5f97/ldaps_in_aio_deployment/

# The idea
When creating the `aio-mastercontainer`, either via `docker run` or via `docker-compose`, it is possible to point the environmental variable `TRUSTED_CACERTS_DIR` to a directory on the host which contains one or more custom CA-Certificates.

If the variable is set, it will be passed to the `aio-nextcloud` container and the given directory will be mounted as a volume under `/usr/local/share/ca-certificates`. Furthermore, the command `update-ca-certificates` will be run in order to make the OS update its trusted certification authorities.

The `aio-nextcloud` container now trusts the given custom CAs. It is now possible, for example, to configure LDAPS against a company domain controller which uses a certificate signed by the company's internal CA.

Trusting custom CAs probably also has other applications / use cases I didn't think about yet.

**Comment by lollo0296** [comment]
@szaimen I had never actually tried out LDAPS adding a CA manually in the nextcloud container, so I just did it.

I usually create a subdirectory named `extra` for my additional CA-Certificates under `/usr/local/share/ca-certificates`, then run `update-ca-certificates`. Although this works in debian based containers, this doesn't seem to be the case for alpine. I was getting some unusual warning and the whole thing wasn't working.

Indeed, as mentioned in https://github.com/gliderlabs/docker-alpine/issues/52#issuecomment-795744705, alpine's `update-ca-certificates` does not like subdirectories and requires putting the additional certificates directly in `/usr/local/share/ca-certificates`. So I changed the mount point accordingly in [php/containers.json](https://github.com/nextcloud/all-in-one/pull/1097/files#diff-717b4bb5f57cb717b22d901996237f21ba634aab886c0afbd2129d3939c883f9)

After doing this test, I can confirm the LDAPS do actually work with our Domain Controller as a server (WinServer2022). Nice.

grafik

**Comment by szaimen** [comment]
@lollo0296 Thanks for your efforts! The PR looks already pretty good! However a few places are missing. 
- [ ] add an entry to the readme.md of this repo
- [ ] add a check to Containers/mastercontainer/start.sh which limits accepted values
- [ ] add a string substitution to manual-install/update-yaml.sh that adds an explanation there, too
- [ ] add the string substition to php/src/Docker/DockerActionManager.php as well
- [ ] add an entry with the setting to tests/QA/060-environmental-variables.md

You can have a look at https://github.com/nextcloud/all-in-one/pull/1028/files for what files need to be adjusted. In each file should already be an example for this kind of value...

**Comment by lollo0296** [comment]
@szaimen So far so good, I implemented all the missing points!

**Comment by szaimen** [comment]
> @szaimen So far so good, I implemented all the missing points!

Thanks! Will have a look soon!

File: Containers/mastercontainer/start.sh [hunk 1]
```diff
@@ -161,6 +161,17 @@ It is set to '$DOCKER_SOCKET_PATH'."
         exit 1
     fi
 fi
+if [ -n "$TRUSTED_CACERTS_DIR" ]; then
+    if ! echo "$TRUSTED_CACERTS_DIR" | grep -q "^/" || echo "$TRUSTED_CACERTS_DIR" | grep -q "/$"; then
+        echo "You've set TRUSTED_CACERTS_DIR but not to an allowed value.
+        It should be an absolute path to a directory that starts and ends with '/'"
```
Comment (szaimen) [inline Containers/mastercontainer/start.sh:167]:
```suggestion
        echo "You've set TRUSTED_CACERTS_DIR but not to an allowed value.
It should be an absolute path to a directory that starts with '/' but not end with '/'.
It is set to '$TRUSTED_CACERTS_DIR '."
```

File: Containers/mastercontainer/start.sh [hunk 2]
```diff
@@ -161,6 +161,17 @@ It is set to '$DOCKER_SOCKET_PATH'."
...
+        exit 1
+    fi
+    if [ ! -d "$TRUSTED_CACERTS_DIR" ]; then
+        echo "You've set TRUSTED_CACERTS_DIR but the given directory does not exist."
+        exit 1
+    ;
```
Comment (szaimen) [inline Containers/mastercontainer/start.sh:173]:
```suggestion
```
This doesn't work from inside of an container, I fear.

File: docker-compose.yml [hunk 3]
```diff
@@ -26,6 +26,7 @@ services:
       # - DISABLE_BACKUP_SECTION=true # Setting this to true allows to hide the backup section in the AIO interface.
       # - NEXTCLOUD_UPLOAD_LIMIT=10G # Can be adjusted if you need more. See https://github.com/nextcloud/all-in-one#how-to-adjust-the-upload-limit-for-nextcloud
       # - NEXTCLOUD_MAX_TIME=3600 # Can be adjusted if you need more. See https://github.com/nextcloud/all-in-one#how-to-adjust-the-max-execution-time-for-nextcloud
+      # - TRUSTED_CACERTS_DIR=/cacerts # CA certificates in this directory will be trusted by the OS of the nexcloud container (Useful e.g. for LDAPS)
```
Comment (szaimen) [inline docker-compose.yml:29]:
```suggestion
      # - TRUSTED_CACERTS_DIR=/path/to/my/cacerts # CA certificates in this directory will be trusted by the OS of the nexcloud container (Useful e.g. for LDAPS) See See https://github.com/nextcloud/all-in-one#how-to-trust-user-defiend-certification-authorities-(ca)
```

File: manual-install/update-yaml.sh [hunk 4]
```diff
@@ -64,6 +64,7 @@ sed -i 's|NEXTCLOUD_DATADIR=|NEXTCLOUD_DATADIR=nextcloud_aio_nextcloud_data
 sed -i 's|NEXTCLOUD_MOUNT=|NEXTCLOUD_MOUNT=/mnt/          # This allows the Nextcloud container to access directories on the host. It must never be equal to the value of NEXTCLOUD_DATADIR!|' sample.conf
 sed -i 's|NEXTCLOUD_UPLOAD_LIMIT=|NEXTCLOUD_UPLOAD_LIMIT=10G          # This allows to change the upload limit of the Nextcloud container|' sample.conf
 sed -i 's|NEXTCLOUD_MAX_TIME=|NEXTCLOUD_MAX_TIME=3600          # This allows to change the upload time limit of the Nextcloud container|' sample.conf
+sed -i 's|TRUSTED_CACERTS_DIR=|TRUSTED_CACERTS_DIR=/path/to/my/directory          # Nextcloud container will trust all the Certification Authorities, whose certificates are included in the given directory.|' sample.conf
```
Comment (szaimen) [inline manual-install/update-yaml.sh:67]:
```suggestion
sed -i 's|TRUSTED_CACERTS_DIR=|TRUSTED_CACERTS_DIR=/path/to/my/cacerts          # Nextcloud container will trust all the Certification Authorities, whose certificates are included in the given directory.|' sample.conf
```

File: tests/QA/060-environmental-variables.md [hunk 5]
```diff
@@ -11,5 +11,7 @@
 - [ ] When starting the mastercontainer with `-e NEXTCLOUD_MAX_TIME=4000` it should change Nextclouds upload max time 4000s. See https://github.com/nextcloud/all-in-one#how-to-adjust-the-max-execution-time-for-nextcloud for allowed values.
 - [ ] When starting the mastercontainer with `-e DOCKER_SOCKET_PATH="/var/run/docker.sock.raw"` it should map `/var/run/docker.sock.raw` to `/var/run/docker.sock` inside the watchtower container which allow to update the mastercontainer on macos and with docker rootless.
 - [ ] When starting the mastercontainer with `-e DISABLE_BACKUP_SECTION=true` it should hide the backup section that gets shown after AIO is set up (everything of [020-backup-and-restore](./020-backup-and-restore.md)) and simply show that the backup section is disabled.
+- [ ] When starting the mastercontainer with `-e TRUSTED_CACERTS_DIR=/path/to/my/folder`, the resulting nextcloud container should trust all the Certification Authorities, whose certificates are included in the directory `/path/to/my/folder/` on the host.
```
Comment (szaimen) [inline tests/QA/060-environmental-variables.md:14]:
```suggestion
- [ ] When starting the mastercontainer with `-e TRUSTED_CACERTS_DIR=/path/to/my/cacerts`, the resulting nextcloud container should trust all the Certification Authorities, whose certificates are included in the directory `/path/to/my/cacerts` on the host.

```

**Review by szaimen — CHANGES_REQUESTED** [review]
Just a few small things to address. Otherwise looks good, thanks! :)

File: Containers/mastercontainer/start.sh (hunk 2 above)
Comment (lollo0296) [reply Containers/mastercontainer/start.sh:173]: That's a point!

**Comment by lollo0296** [comment]
> Just a few small things to address. Otherwise looks good, thanks! :)

All done ✅

**Comment by szaimen** [comment]
Thanks! Can you please squash your changes into one commit?

**Comment by lollo0296** [comment]
@szaimen I really don't know how to do that. I made 13 commits, but when I run `git reset --soft HEAD~13` than I suddenly have changes from other people in there, and it is a real mess to cherry-pick out my changes. 

Can't you just `git merge --squash` ? It should works like a charm that way

**Comment by szaimen** [comment]
@lollo0296 I'll try to squash and rebase for you. Can you please allow me to push to your branch?

**Comment by lollo0296** [comment]
@szaimen I invited you as a collaborator in my fork :)

**Review by szaimen — APPROVED** [review]
lgtm now, thanks! :)

**Comment by lollo0296** [comment]
Nice @szaimen, I can't wait to try it out and see if LDAPS will now work out of the box when starting the AIO-Stack 😍

**Comment by szaimen** [comment]
This is now released with v2.0.3 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Pull request merged**